# HG changeset patch
# User Avinoam Kalma <a.kalma@gmail.com>
# Date 1485808958 -3600
#      Mon Jan 30 21:42:38 2017 +0100
# Branch stable
# Node ID 46b914a213db939d750acead908d717427c0d694
# Parent  8cb4a2c07bce14a14475c2e969e8caf709c3476d
Workaround segfault when an error occurs while printing (bug #49779)

* print.m: check early that the requested directory exists and the requested
  file is writable
* __print_parse_opts__.m: error out if the specified device/extension is unknown

diff --git a/scripts/plot/util/print.m b/scripts/plot/util/print.m
--- a/scripts/plot/util/print.m
+++ b/scripts/plot/util/print.m
@@ -309,16 +309,28 @@
 ##
 ## @seealso{saveas, hgsave, orient, figure}
 ## @end deftypefn
 
 function print (varargin)
 
   opts = __print_parse_opts__ (varargin{:});
 
+  folder = fileparts (opts.name);
+  if (! isempty (folder) && ! exist (folder,"dir"))
+    error ("print: directory %s does not exist", folder);
+  endif
+
+  fid = fopen (opts.name, "w+");
+  if (fid == -1)
+    error ("print: cannot open file %s for writing", opts.name);
+  endif
+  fclose (fid);
+  unlink (opts.name);
+  
   opts.pstoedit_cmd = @pstoedit;
   opts.fig2dev_cmd = @fig2dev;
   opts.latex_standalone = @latex_standalone;
   opts.lpr_cmd = @lpr;
   opts.epstool_cmd = @epstool;
 
   if (isempty (opts.figure) || ! isfigure (opts.figure))
     error ("print: no figure to print");
diff --git a/scripts/plot/util/private/__print_parse_opts__.m b/scripts/plot/util/private/__print_parse_opts__.m
--- a/scripts/plot/util/private/__print_parse_opts__.m
+++ b/scripts/plot/util/private/__print_parse_opts__.m
@@ -257,17 +257,17 @@ function arg_st = __print_parse_opts__ (
   if (strcmp (__graphics_toolkit__, "gnuplot"))
     suffixes(strncmp (dev_list, "epslatex", 8)) = {"eps"};
   endif
 
   match = strcmpi (dev_list, arg_st.devopt);
   if (any (match))
     default_suffix = suffixes{match};
   else
-    default_suffix = arg_st.devopt;
+    error ("print: unknown device %s", arg_st.devopt);  
   endif
 
   if (dot == 0 && ! isempty (arg_st.name))
     arg_st.name = [arg_st.name "." default_suffix];
   endif
 
   if (arg_st.append_to_file)
     if (isempty (arg_st.name))
